pep-9999.rst

```diff
@@ -0,0 +1,1102 @@
+`here <https://erictraut/cpython/typeparam>`_.
+
+
+Rejected Ideas
+==============
+
+Prefix Clause
+-------------
+We explored various syntactic options for specifying type parameters that
+preceded ``def`` and ``class`` statements. One such variant we considered
+used a ``using`` clause as follows:
+
+::
+
+    using S, T
+    class ClassA: ...
+
+This option was rejected because the scoping rules for the type parameters
+were less clear. Also, this syntax did not interact well with class and
+function decorators, which are common in Python. Only one other popular
+programming language, C++, uses this approach.
+
+
+Extends Keyword
+---------------
+We considered introducing an "extends" keyword for specifying the upper bound
+of a type parameter. This is consistent with Java and TypeScript, two other
+popular languages. However, "extends" didn't feel very Pythonic. It also didn't
+lend itself to supporting constrained type parameters.
+
+
+Explicit Variance
+-----------------
+We considered adding syntax for specifying whether a type parameter is intended
+to be invariant, covariant, or contravariant. The ``typing.TypeVar`` mechanism
+in Python requires this. A few other languages including Scala and C# provide
+also require developers to specify the variance. We rejected this idea because
+variance can generally be inferred, and most modern programming languages
+do infer variance based on usage. Variance is an advanced topic that
+many developers find confusing, so we want to eliminate the need to 
+understand this concept for most Python developers.
+
+
+Name Mangling
+-------------
+When considering implementation options, we considered a "name mangling"
+approach where each type parameter was given a unique "mangled" name by the
+compiler. This mangled name would be based on the qualified name of the
+generic class, function or type alias it was associated with. This approach
+was rejected because qualified names are not necessarily unique, which means
+the mangled name would need to be based on some other randomized value.
+
+
+Appendix A: Survey of Type Parameter Syntax
+===========================================
+
+Support for generic types is found in many programming languages. In this
+section, we provide a survey of the options used by other popular programming
+languages. This is relevant because familiarity with other languages will
+make it easier for Python developers to understand this concept.
+
+C++
+---
+
+C++ uses angle brackets in combination with keywords "template" and
+"typename" to declare type parameters. It uses angle brackets for
+specialization.
+
+C++20 introduced the notion of generalized constraints, which can act
+like protocols in Python. A collection of constraints can be defined in
+a named entity called a "concept".
+
+Variance is not explicitly specified, but constraints can enforce variance.
+
+A default type argument can be specified using the "=" operator.
```

Comment:
Is this a holdover from earlier versions of the PEP, in which the PEP also proposed a mechanism to specify a default type argument? It seems a bit irrelevant to the stuff the rest of the PEP is discussing.

Reply:
No, this isn't a holdover. I included this information as part of a general survey. It will be useful in informing future typing PEPs such as [this one](https://gist.github.com/Gobot1234/8c9bfe8eb88f5ad42bf69b6f118033a7).

Reply:
Okay — at the moment it doesn't seem obvious to me why this information is included here, since it touches on a point that hasn't hitherto been discussed in the PEP.